extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,236 @@
+   * @param {!Object=} opt_event Object with details about the event.
+   * @param {number=} opt_iterations Number of recursive expansions to perform.
+   *    Defaults to 2 substitutions.
+   * @param {boolean=} opt_encode Used to determine if the vars should be
+   *    encoded or not. Defaults to true.
+   * @return {string} The expanded string.
+   */
+  expandTemplate(template, trigger, config, opt_event, opt_iterations,
+      opt_encode) {
+    opt_iterations = opt_iterations === undefined ? 2 : opt_iterations;
+    opt_encode = opt_encode === undefined ? true : opt_encode;
+    if (opt_iterations < 0) {
+      user().error('Maximum depth reached while expanding variables. Please ' +
+          'ensure that the variables are not recursive.');
+      return template;
+    }
+
+    // Replace placeholders with URI encoded values.
+    // Precedence is opt_event.vars > trigger.vars > config.vars.
+    // Nested expansion not supported.
+    return expandTemplate(template, key => {
+      // TODO (avimehta): The parsing for variables breaks when `|` is used as
+      // something other than the filter separator. A full-fledged parser would
+      // be needed to fix this.
+
+      const tokens = key.split('|');
+      if (!tokens[0]) {
+        return '';
+      }
+
+      const {name, argList} = this.getNameArgs_(tokens[0].trim());
+      let raw = (opt_event && opt_event['vars'] && opt_event['vars'][name]) ||
+          (trigger['vars'] && trigger['vars'][name]) ||
+          (config['vars'] && config['vars'][name]) ||
+          '';
+
+      // Values can also be arrays and objects. Don't expand them.
+      if (typeof raw == 'string') {
+        raw = this.expandTemplate(raw, trigger, config, opt_event,
+          opt_iterations - 1);
+      }
+
+      raw = this.applyFilters_(raw, tokens.slice(1));
+      const val = opt_encode ? this.encodeVars(raw, name) : raw;
+      return val ? val + argList : val;
+    });
+  }
+
+  /**
+   * Returns an array containing two values: name and args parsed from the key.
+   *
+   * @param {string} key The key to be parsed.
+   * @return {!Object<string>}
+   * @private
+   */
+  getNameArgs_(key) {
+    if (!key) {
+      return {name: '', argList: ''};
+    }
+    const match = key.match(/([^(]*)(\([^)]*\))?/);
+    if (!match) {
+      user().error(TAG, 'Variable with invalid format found: ' + key);
+    }
+    return {name: match[1], argList: match[2] || ''};
+  }
+
+  /**
+   * @param {string|!Array<string>} raw The values to URI encode.
+   * @param {string} unusedName Name of the variable.
+   * @return {string} The encoded value.
+   */
+  encodeVars(raw, unusedName) {
+    if (!raw) {
+      return '';
+    }
+
+    if (isArray(raw)) {
+      return raw.map(encodeURIComponent).join(',');
+    }
+    // Separate out names and arguments from the value and encode the value.
+    const {name, argList} = this.getNameArgs_(String(raw));
+    return encodeURIComponent(name) + argList;
```

Comment:
still not addressed. I am aware of this and have not had the time to go through all the possible scenarios here.